### DISCUSSION TITLE
FM Synth Expansion

### DISCUSSION THREAD

**Discussion by m-m-adams** [SynthstromAudible/DelugeFirmware#227] [Ideas]
Putting the points from #130 here, I was planning to work on it when I opened the issue but I'll be spending a bunch more time on filters first

- More waveforms
    - The FM synth could be expanded to allow selecting between waveforms rather than just sine. In order to do this without aliasing the maximum frequency generated by the underlying oscillator must be able to be configured dynamically, using a method such as blep/blit.
    - The maximum frequency cutoffs for an oscillator to avoid aliasing can be calculated by inverting the methods in chownings 1973 paper.
    - Some filters would have to be enabled for this to be useful - the SVF or a simple LPF without drive would minimize the impact on voice count
- More algorithms
    - Easiest is expanding the destination menu to allowing both modulation oscillators to be routed to either or both carriers. I think this fits best with deluge's workflow
    - More complicated is adding an algorithm selector, possibly on the pad above 'destination', for 4-op FM algorithms
    - Most complicated is adding a user algorithm screen. Pressing destination could open a new view on the upper left four pads, with destinations 1-4 at the top and sources 1-4 on the left. Pressing a pad would let you send a source to a destination
- More envelopes 
    - Having more things to modulate makes more envelopes useful. A way to show 2 more envelopes would be nice

**Comment by PaulFreund** [comment]
Overall sounds like a really good plan. So far I never managed to create reproducible results from FM synths, only happy little accidents so I can't really give an educated suggestio here. Only thing I can tell is that having a list of predefined algorithm would probably most familiar but also harder to integrate with deluges workflow and the question how many operators/modulators would be feasible also might have an impact if more predefined algorithms even make sense. I'd go with the first option first and then see how well it works

**Comment by weavermedia** [comment]
I spent a long time with the FM synth engine, alongside my Yamaha FB-01, TX81Z, DX-100, Reface DX and Korg FM2. I made a FM synth pack which was received quite well. https://meatbeats.gumroad.com/l/freq-mode-deluge. If anyone working on the FM stuff wants a free download code let me know.

I'd like to see some compatibility with the vast amount of existing FM sysex presets out there. Which means moving towards some kind of feature parity. But even with the sorta kinda 2-operator algorithm on the Deluge I was able to reproduce many so-called 4-operator sounds simply because in many cases 2 of those operators were doing what the Deluge can do with its own unique routing.

More waveforms - some vintage FM synths have extra waveforms although they're really only rectified, or otherwise modified, sine waves. But having those 'extra' waveforms would be a good start towards sysex compatibility.

More algorithms - I feel like replicating the existing classic 4-operator algorithms would be a great start, them expand from that. A list would make the most sense to me. Showing the algorithm in the UI is a challenge for sure. But a well designed section in the manual will help with that.

Vintage FM synths have a lot of seemingly arbitrary limitations. Brighter minds than mine will certainly think of exciting new ways to route signals and come up with brand new sounds.

In the interim would it make sense for me to make a list comparing vintage 4-operator capabilities with the current Deluge system? Or should I abandon all thinking around getting vintage sysex compatibility?

**Reply by m-m-adams** [reply]
I think sysex compatibility for the deluge FM synth is probably the wrong way to go - Id like to keep its focus on the easy on device editing and just slightly expand its sound design capabilities to be a hybrid FM+subtractive engine

Sysex support in the seperate DX7 oscillator type that's being worked on would make a lot of sense though!

**Reply by weavermedia** [reply]
If the DX7 oscillator type becomes a reality that would solve both problems 👍

**Comment by fenixsong** [comment]
I had a similar idea so I wanted to jump on this thread. I like what you said about "More algorithms
Easiest is expanding the destination menu to allowing both modulation oscillators to be routed to either or both carriers. I think this fits best with deluge's workflow".  I think this would be the cleanest way to go and wouldn't take up any more "buttons". As this could get quite involved to get many new types of algorithms, I thought of a few easy wins as a first step:
  
Stick to the "4 OP" that deluge has now. But 4 envelops would really help it become real 4OP
If you look at the FM Structure Diagrams of Deluge you see that if there are 2 cariers and 2 modulators with the modulaters going to both carriers you have no choice but to have each of the modulators modulating each carrier. What would be an easy addition is to add a separate "algorithm" that just separates them and have 1 modulator only modulating 1 carrier, and then again the other modulator only modulating the 2nd carrier. I like the cross modulation but we could make many more synth designs if there was separate pairs.

Beyond that it becomes a little more tricky but there are a few more easy additions:
could there be a 4 op chain like: Mod->Mod->Mod->Car ? This could do a lot of classic FM tones but maybe not the most needed based on Deluge structure (because of all the feedback opportunities). This may need more restructuring of the FM architecture but idk how hard it would be.
Could there be 3 Modulators all going into one Carrier?
Could all 4 ops be Carriers?

In another area: could modulators become fixed frequency, like on DX and others? Hmmm, is there already a way to do this?

Hope these ideas help. Let me know if I can help. I looked at the code but don't know if I am ready to hack the fm synth yet.

**Reply by m-m-adams** [reply]
With regards to algorithms I was thinking of just letting both modulators choose a destination of either/both carriers or the other modulator. It wouldn't let you make a 4 op stack but imo that's the least useful 4 op algorithm anyway

**Comment by sissyfus** [comment]
More than a 4-op I would wish for opening up to more waveforms/wavetables and also samples as operators. I do not know how technically difficult this is to achieve but it would make the Deluge able to do something very few FM synths can.

**Comment by yratof** [comment]
Has anyone suggested the idea of porting over the Mutable Instruments engines from Braids or Plaits? 

https://github.com/pichenettes/eurorack/blob/08460a69a7e1f7a81c5a2abcc7189c9a6b7208d4/braids/macro_oscillator.h

Would help with the 'thin' sounds that everyone points at

**Reply by yratof** [reply]
ie: https://github.com/pichenettes/eurorack/tree/master/plaits/dsp
